touchforms/formplayer/experiments.py

```diff
@@ -0,0 +1,102 @@
+import laboratory
+import json
+import logging
+
+diff_logger = logging.getLogger('formplayer_diff')
+timing_logger = logging.getLogger('formplayer_timing')
+
+
+class FormplayerExperiment(laboratory.Experiment):
+    session_id_mapping = {}
+
+    def publish(self, result):
+        # if we're starting a new form, we need to store the mapping between session_ids so we can use later
+        if (self.name == "new-form"):
+            control_session_id = json.loads(result.control.value)["session_id"]
+            candidate_session_id = json.loads(result.observations[0].value)["session_id"]
+            FormplayerExperiment.session_id_mapping[control_session_id] = candidate_session_id
+
+        control = result.control
+        # We're only ever returning one of these (I think)
+        candidate = result.observations[0]
+
+        self.log_timing(control, candidate)
+
+        control_value = json.loads(result.control.value)
+        candidate_value = json.loads(result.observations[0].value)
+
+        if not formplayer_compare(control_value, candidate_value):
+            self.log_diff(control_value, candidate_value)
+
+    def log_diff(self, control_value, candidate_value):
+        request = self.context['request']
+        action = request['action']
+        diff_logger.info(
+            extra={
+                'action': action,
+                'request': request,
+                'control': control_value,
+                'candidate': candidate_value
+            })
+
+    def log_timing(self, control, candidate):
+        timing_logger.info(
+            "MSG",
+            extra={
+                'action': self.name,
+                'control_duration': control.duration,
+                'candidate_duration': candidate.duration
+            })
+
+
+def compare_list(control, candidate):
```

Comment:
would be good to also check that candidate is a list before doing the rest of the checks

Reply:
nevermind, saw you added this later